internal/controllers/configuration/kongupstreampolicy_utils.go

```diff
@@ -0,0 +1,178 @@
+package configuration
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gatewaycontroller "github.com/kong/kubernetes-ingress-controller/v3/internal/controllers/gateway"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/gatewayapi"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1beta1"
+)
+
+type serviceStatus struct {
+	service           corev1.Service
+	acceptedCondition metav1.Condition
+}
+
+// -----------------------------------------------------------------------------
+// KongUpstreamPolicy Controller - Reconciler Helpers
+// -----------------------------------------------------------------------------
+
+// enforceKongUpstreamPolicyStatus gets a list of services (ancestors) along with their desired status and enforce them
+// in the KongUpstreamPolicy status.
+func (r *KongUpstreamPolicyReconciler) enforceKongUpstreamPolicyStatus(ctx context.Context, oldPolicy *kongv1beta1.KongUpstreamPolicy, servicesStatus []serviceStatus) (bool, error) {
+	newPolicyStatus := gatewayapi.Policystatus{}
+	if len(servicesStatus) > 0 {
+		newPolicyStatus.Ancestors = make([]gatewayapi.PolicyAncestorStatus, 0, len(servicesStatus))
+	}
+	for _, ss := range servicesStatus {
+		newPolicyStatus.Ancestors = append(newPolicyStatus.Ancestors,
```

Comment:
The `Ancestors` of `PolicyStatus` in `gatewayapiv1alpha2` has the limit of `MaxItem=16`:
 https://github.com/kubernetes-sigs/gateway-api/blob/a6c0e9ce5dd38cc3a065a672d4132f70a63cf087/apis/v1alpha2/policy_types.go#L218. We should also limit the length.

Reply:
Yep, this limit is a problem. We should decide how to proceed in case more than 16 services reference the same upstream. The options that I see:
1. introduce service validation in our validating webhook. If the 17th service references the same policy, the update/creation gets rejected.
2. simply log/emit an event from the `KongUpstreamPolicy` controller
3. add a new condition in the services that breaks this constraint

I lean toward n. 2, but I'm curious to hear your thoughts about this (even if you have other options other than the 3 I listed here) @randmonkey @czeslavo

Reply:
Yeah, 2nd option looks the most appealing to me. I think it's more important to still allow attachment to more than 16 Services than providing conditions. We could emit an event for `KongUpstreamPolicy` that would say something like "Successfully accepted by Service %s" (BTW, this could be emitted despite the number of conditions to not make it more complicated).

Reply:
Ok, so you are suggesting we allow references for as many objects as the user wants, but we limit the number of ancestors in the status to 16. Correct? I do not like that much this approach, but actually we have no alternative 🤔

Reply:
> We could emit an event for KongUpstreamPolicy that would say something like "Successfully accepted by Service %s" (BTW, this could be emitted despite the number of conditions to not make it more complicated).

I don't know, emitting an event to say that status has been updated with some information looks like a duplicated feedback to me. I think that status update is enough, along with logging about ancestors not settable in the status due to the high number of referring services.